shared/engine/index.js

```diff
@@ -1,421 +1,257 @@
-  _failOnError: boolean;
-  inListenerCallback: boolean;
+  // Tracking outstanding sessions
+  _sessionsMap: {[key: SessionIDKey]: Session} = {}
+  // Helper we delegate actual calls to
+  _rpcClient: createClientType
+  // All incoming call handlers
+  _incomingHandler: {[key: MethodKey]: (param: Object, response: ?Object) => void} = {}
+  // Keyed methods that care when we reconnect
+  _onConnectHandlers: {[key: string]: () => void} = {}
+  // Set to true to throw on errors. Used in testing
+  _failOnError: boolean = false
+  // We generate sessionIDs monotonically
+  _nextSessionID: number = 123
 
   constructor () {
-    const {logLocal} = setupLocalLogs()
-    this.logLocal = logLocal
-
-    // Keep some meta data from session ID to response meta
-    // To help debug outstanding requests
-    this.responseMeta = {}
-
-    if (printOutstandingRPCs) {
-      setInterval(() => {
-        const keys = Object.keys(this.sessionIDToResponse).filter(k => this.sessionIDToResponse[k])
-        if (keys.length) {
-          keys.forEach(k => {
-            const {method, param} = this.responseMeta[k]
-            logLocal('Outstanding RPC sessionIDs: %s.\nMethod: %s with param: %O', k, method, param)
-          })
-        }
-      }, 10 * 1000)
-    }
-
-    if (!NO_ENGINE) {
-      this.rpcClient = createClient(
-        payload => {
-          if (__DEV__ && KEYBASE_RPC_DELAY_RESULT) {
-            setTimeout(() => this._rpcIncoming(payload), KEYBASE_RPC_DELAY_RESULT)
-          } else {
-            this._rpcIncoming(payload)
-          }
-        },
-        () => this._onConnect()
-      )
-    }
-
-    this.sessionID = 123
-
-    // to find callMap for rpc callbacks, including waitingHandler
-    this.sessionIDToIncomingCall = {}
-
-    // any non-responded to response objects
-    this.sessionIDToResponse = {}
-
-    // seqid to wrapped response so we can handle cancelations
-    this.seqIDToWrappedResponse = {}
-
-    // A call map for general listeners
-    // These are commands that the service can call at any point in time
-    this.generalListeners = {}
+    this._rpcClient = createClient(
+      payload => this._rpcIncoming(payload),
+      () => this._onConnected()
+    )
 
-    this.serverListeners = {}
-
-    // A list of functions to call when we connect
-    this.onConnectFns = {}
-
-    // Throw an error and fail?
-    this._failOnError = false
+    this._setupCoreHandlers()
+    this._setupDebugging()
   }
 
-  _onConnect () {
-    this.inListenerCallback = true
-    Object.keys(this.onConnectFns).forEach(k => this.onConnectFns[k]())
-    this.inListenerCallback = false
-  }
-
-  listenOnConnect (key: string, f: () => void) {
-    if (NO_ENGINE) {
+  _setupDebugging () {
+    if (!__DEV__) {
       return
     }
 
-    if (!f) {
-      throw new Error('Null callback sent to listenOnConnect')
-    }
-
-    if (this.inListenerCallback) {
-      throw new Error('Ran a listenOnConnect within another listenOnConnect')
```

Comment:
What happened with this?

This is pretty useful, because we don't want to allow this. (I ran into this recently)